sycl/test-e2e/Graph/Inputs/spec_constants_handler_api.cpp

```diff
@@ -0,0 +1,206 @@
+
+constexpr sycl::specialization_id<int> IntId;
+constexpr sycl::specialization_id<int> IntId2(2);
+constexpr sycl::specialization_id<float> FloatId(3.14);
+
+class TestDefaultValuesKernel;
+class EmptyKernel;
+class TestSetAndGetOnDevice;
+
+bool test_default_values(sycl::queue Queue);
+bool test_set_and_get_on_host(sycl::queue Queue);
+bool test_set_and_get_on_device(sycl::queue Queue);
+
+bool test_set_and_get_on_device(sycl::queue Queue);
+
+int main() {
+  auto ExceptionHandler = [&](sycl::exception_list Exceptions) {
+    for (std::exception_ptr const &E : Exceptions) {
+      try {
+        std::rethrow_exception(E);
+      } catch (sycl::exception const &E) {
+        std::cout << "An async SYCL exception was caught: " << E.what()
+                  << std::endl;
+        std::exit(1);
+      }
+    }
+  };
+
+  queue Queue{ExceptionHandler,
+              {sycl::ext::intel::property::queue::no_immediate_command_list{}}};
+
+  if (!test_default_values(Queue)) {
+    std::cout << "Test for default values of specialization constants failed!"
+              << std::endl;
+    return 1;
+  }
+
+  if (!test_set_and_get_on_host(Queue)) {
+    std::cout << "Test for set and get API on host failed!" << std::endl;
+    return 1;
+  }
+
+  if (!test_set_and_get_on_device(Queue)) {
+    std::cout << "Test for set and get API on device failed!" << std::endl;
+    return 1;
+  }
+
+  return 0;
+};
+
+bool test_default_values(sycl::queue Queue) {
+  sycl::buffer<int> IntBuffer(1);
+  IntBuffer.set_write_back(false);
+  sycl::buffer<int> IntBuffer2(1);
+  IntBuffer2.set_write_back(false);
+  sycl::buffer<float> FloatBuffer(1);
+  FloatBuffer.set_write_back(false);
+
+  {
+    exp_ext::command_graph Graph{
+        Queue.get_context(),
+        Queue.get_device(),
+        {exp_ext::property::graph::assume_buffer_outlives_graph{}}};
+
+    add_node(Graph, Queue, ([&](sycl::handler &CGH) {
+               auto IntAcc =
+                   IntBuffer.get_access<sycl::access::mode::write>(CGH);
+               auto IntAcc2 =
+                   IntBuffer2.get_access<sycl::access::mode::write>(CGH);
+               auto FloatAcc =
+                   FloatBuffer.get_access<sycl::access::mode::write>(CGH);
+
+               CGH.single_task<TestDefaultValuesKernel>(
+                   [=](sycl::kernel_handler KH) {
+                     IntAcc[0] = KH.get_specialization_constant<IntId>();
+                     IntAcc2[0] = KH.get_specialization_constant<IntId2>();
+                     FloatAcc[0] = KH.get_specialization_constant<FloatId>();
+                   });
+             }));
+
+    auto GraphExec = Graph.finalize();
+
+    Queue.submit([&](handler &CGH) { CGH.ext_oneapi_graph(GraphExec); });
+    Queue.wait_and_throw();
+  }
+
+  sycl::host_accessor IntAcc(IntBuffer, sycl::read_only);
+  if (!check_value(
+          0, IntAcc[0],
+          "integer specialization constant (defined without default value)"))
+    return false;
```

Comment:
Might it be useful to let the other test cases run and fail too, just to have as much information as possible or are they too closely tied that it won't give much value?

Reply:
Good idea, updated to try the following tests rather than exiting early.